# Adding County Details to Database

*For context view the ReadMe here:* https://github.com/amcgaha/camp-covid-geography/blob/main/README.md 

This script bolsters an existing database I created to store information on our campers and their households. Before adding these extra data, the only address data present in the database were household mailing addresses.

Here we add necessary tables to convert zip codes to county FIPS and county names. This will allow us to compare COVID data at the county level with our campers' zip codes. We also add county population data, which we will later use to calculate the case rate per 100,000 people in each county.

The existing database is stored using PostgreSQL on a local machine. We will use SQLalchemy to interact with it.  

In [1]:
import pandas as pd
import requests
import io
import psycopg2
from sqlalchemy import create_engine, MetaData, Table, Column, String, Integer, Numeric, Float

## 1. Zip Code to Counties

Source: https://www.huduser.gov/portal/datasets/usps_crosswalk.html#codebook

We need to import and validate this data before we can add it to our database. The first step is downloading the csv from GitHub and reading it to a dataframe. Then we print out the first five rows to see what we've got.

In [2]:
zip_to_county_url = 'https://raw.githubusercontent.com/amcgaha/camp-covid-geography/main/zip_to_county_hud.csv'

download_zip = requests.get(zip_to_county_url).content

zip_df = pd.read_csv(io.StringIO(download_zip.decode('utf-8')))

print(zip_df.head())

   ZIP  COUNTY  RES_RATIO  BUS_RATIO  OTH_RATIO  TOT_RATIO
0  501   36103   0.000000   1.000000   0.000000   1.000000
1  601   72113   0.160485   0.197044   0.128834   0.162013
2  601   72001   0.839515   0.802956   0.871166   0.837987
3  602   72005   0.000000   0.001202   0.000000   0.000081
4  602   72003   1.000000   0.998798   1.000000   0.999919


We can see that we have some other columns in addition to the zip codes and county FIPS. By reading the source website, we know that these columns indicate the ratio of residential, business, and other addresses in the zip code. While we don't need this data now, it might be interesting to study later. Let's keep it. 

The first clear problem that the data present is with the zip codes. We know that zip codes have five digits, right? The first rows, at least, have only three. A Google search indicates that these may be prefixes representing larger areas. However, a search for zip code 501 at the USPS website returns no match. (https://m.usps.com/m/ZipLookupAction?search=zip)

Another possibility is that there's something wrong with the dataset. Let's make sure we have valid zip codes here. 

To start, let's sort the values from highest to lowest and see what the highest zip codes look like.

In [3]:
print(zip_df.sort_values('ZIP', ascending=False).head())

         ZIP  COUNTY  RES_RATIO  BUS_RATIO  OTH_RATIO  TOT_RATIO
54192  99929    2275        0.0        0.0        1.0        1.0
54191  99928    2130        0.0        0.0        1.0        1.0
54190  99927    2198        0.0        0.0        1.0        1.0
54189  99926    2198        0.0        0.0        1.0        1.0
54188  99925    2198        0.0        0.0        1.0        1.0


So we do have five-digit codes in here. A quick lookup of the top zip code tells us we do have a real zip code, too. The first one is from Wrangell, Alaska. 

To understand what's going on -- and to ensure we have the right data -- it may help to get a sense of how many five-digit and three-digit codes we're working with. Let's count the values in each category now.

In [4]:
three_digit_df = zip_df[zip_df['ZIP'] < 1000]
four_digit_df = zip_df[(zip_df['ZIP'] >= 1000) & (zip_df['ZIP'] < 10000)]
five_digit_df = zip_df[zip_df['ZIP'] <= 10000]

three_digit_count = three_digit_df['ZIP'].count()
four_digit_count = four_digit_df['ZIP'].count()
five_digit_count = five_digit_df['ZIP'].count()

print('Three digits: ' + str(three_digit_count))
print('Four digits: ' + str(four_digit_count))
print('Five digits: ' + str(five_digit_count))

Three digits: 332
Four digits: 3313
Five digits: 3645


Uh oh! Now we have a ton of four-digit zip codes too. Consulting the data documentation page, we see that the zip code column is indeed supposed to only contain "5 digit USPS ZIP code." So what's going on here? 

You may have already guessed. A bit of research into zip codes -- or having spent time living in / writing postcards to the northeast -- will tell you that some zip codes begin with 0 or 00. This may be the answer! 

We may have made a mistake by not specifying the data type of each column. Pandas may have interpreted the zip codes as integers, causing it to drop the leading zeroes. Let's see what happens when we specify the correct data types.

In [5]:
proper_dtypes = {'ZIP': 'str', 
                 'COUNTY': 'str', 
                 'RES_RATIO': 'float', 
                 'BUS_RATIO': 'float', 
                 'OTH_RATIO': 'float', 
                 'TOT_RATIO': 'float'}

zip_fixed_df = pd.read_csv(io.StringIO(download_zip.decode('utf-8')), dtype=proper_dtypes)

print(zip_fixed_df.head())

     ZIP COUNTY  RES_RATIO  BUS_RATIO  OTH_RATIO  TOT_RATIO
0  00501  36103   0.000000   1.000000   0.000000   1.000000
1  00601  72113   0.160485   0.197044   0.128834   0.162013
2  00601  72001   0.839515   0.802956   0.871166   0.837987
3  00602  72005   0.000000   0.001202   0.000000   0.000081
4  00602  72003   1.000000   0.998798   1.000000   0.999919


Yes. It looks like pandas had dropped the leading zeroes. We should make sure we specify zip codes as strings in the future, particularly when we're uploading this table to our database. 

Final check - let's look up the first zip code again and see what we get:

Confirmed! A search shows the first zip code is in Holtsville, New York.

__What's next?__

We should also validate the county FIPS. Our next table will provide county FIPS and names, so we can use that table to validate this one. Let's move on for now.

## 2. County Population

Source: https://usafacts.org/visualizations/coronavirus-covid-19-spread-map/

In [6]:
county_pop_url = 'https://raw.githubusercontent.com/amcgaha/camp-covid-geography/main/covid_county_population_usafacts.csv'
    
download_pop = requests.get(county_pop_url).content

# This time let's be sure to specify data types!
population_dtypes = {'countyFIPS': 'str',
                     'County Name': 'str',
                     'State': 'str',
                     'population': 'int'} 
pop_df = pd.read_csv(io.StringIO(download_pop.decode('utf-8')), dtype=population_dtypes)

print(pop_df.head())

  countyFIPS            County Name State  population
0          0  Statewide Unallocated    AL           0
1       1001         Autauga County    AL       55869
2       1003         Baldwin County    AL      223234
3       1005         Barbour County    AL       24686
4       1007            Bibb County    AL       22394


Right away we can see an issue. The table includes at least one row that could pose problems, the row with 'Statewide Unallocated' population. Let's see how many counties have a FIPS of zero, a county name 'Statewide Unallocated', or a population of '0'. 

In [7]:
zero_fips_df = pop_df[pop_df['countyFIPS'] == '0']
statewide_df = pop_df[pop_df['County Name'] == 'Statewide Unallocated']
zero_pop_df = pop_df[pop_df['population'] < 1]

zero_fips_count = zero_fips_df['countyFIPS'].count()
statewide_count = statewide_df['County Name'].count()
zero_pop_count = zero_pop_df['population'].count()

print('Zero FIPS: ' + str(zero_fips_count))
print('Statewide U.A.: ' + str(statewide_count))
print('Zero Pop: ' + str(zero_pop_count))

Zero FIPS: 51
Statewide U.A.: 50
Zero Pop: 53


Most likely we have one 'Statewide Unallocated' for each state. Let's print the head of the subset dataframe to make sure.

In [8]:
print(statewide_df.sort_values('State').head(10))

    countyFIPS            County Name State  population
68           0  Statewide Unallocated    AK           0
0            0  Statewide Unallocated    AL           0
115          0  Statewide Unallocated    AR           0
99           0  Statewide Unallocated    AZ           0
191          0  Statewide Unallocated    CA           0
251          0  Statewide Unallocated    CO           0
316          0  Statewide Unallocated    CT           0
325          0  Statewide Unallocated    DE           0
330          0  Statewide Unallocated    FL           0
398          0  Statewide Unallocated    GA           0


Looks like there is no useful information here, so we can safely get rid of it. Let's create a new dataframe without this county name.

In [9]:
pop_no_state_df = pop_df[pop_df['County Name'] != 'Statewide Unallocated']

print(pop_no_state_df.head())

  countyFIPS     County Name State  population
1       1001  Autauga County    AL       55869
2       1003  Baldwin County    AL      223234
3       1005  Barbour County    AL       24686
4       1007     Bibb County    AL       22394
5       1009   Blount County    AL       57826


We removed 50 duds, but we should still have one blank FIPS and three rows with no population. Let's focus in on those rows. 

In [10]:
print(pop_no_state_df[pop_no_state_df['countyFIPS'] == '0'])

print('==============================================================')

print(pop_no_state_df[pop_no_state_df['population'] < 1])

     countyFIPS                County Name State  population
1862          0  New York City Unallocated    NY           0
     countyFIPS                 County Name State  population
95         2270    Wade Hampton Census Area    AK           0
192        6000  Grand Princess Cruise Ship    CA           0
1862          0   New York City Unallocated    NY           0


Interesting! We can get rid of the NYC Unallocated, but what about the others? 

The Grand Princess Cruise Ship might seem out of place, but it's in here because this data comes from a COVID-19 tracking page. Because we are adding this to a database on camper households, it wouldn't be useful to include a temporary location like a cruise ship. Let's remove it.

That leaves the Wade Hampton Census Area. Some research shows that this is the old name of a census area in Alaska, now called the Kusilvak Census Area. Knowing the vast frontiers of Alaska, this might be a place with zero population. Let's check. 

Uh oh. According to Wikipedia, the population in the 2010 census was 7,459. We have a bad data point!

Before we try to replace it, let's see if the new name exists with the correct population.

In [11]:
print(pop_no_state_df[pop_no_state_df['County Name'].str.contains('Kusilvak')])

   countyFIPS           County Name State  population
83       2158  Kusilvak Census Area    AK        8314


There it is. Notice that the county FIPS changed as well as the name. Perhaps it was included in the table by accident. If we wanted more information we could contact the source or conduct more research. I know that we have no campers from Alaska, and so I feel comfortable with leaving it out. 

Let's create a final dataframe without any of these unhelpful values.

In [12]:
pop_final_df = pop_no_state_df[pop_no_state_df['population'] != 0]

## 3. Matching Zip Code to County

The last step before we store in the database is to merge the tables together. This will also help us validate that our county codes are correct in the zip code dataframe. We will join the dataframes on county FIPS, and then we will pick a few known zip codes and try to return the corresponding county.

Recall that our zip code dataframe is called zip_fixed_df, and we need to join it on 'COUNTY'.

Since our camper households in the database only have zip codes, we only care about counties that match with zip codes. Therefore our join will be an inner join.

In [13]:
joined_df = pd.merge(zip_fixed_df, pop_final_df, left_on='COUNTY', right_on='countyFIPS', how='inner')

print(joined_df.head())

     ZIP COUNTY  RES_RATIO  BUS_RATIO  OTH_RATIO  TOT_RATIO countyFIPS  \
0  00501  36103   0.000000        1.0        0.0   1.000000      36103   
1  06390  36103   0.000000        1.0        1.0   1.000000      36103   
2  11701  36103   0.994713        1.0        1.0   0.995577      36103   
3  11702  36103   1.000000        1.0        1.0   1.000000      36103   
4  11703  36103   1.000000        1.0        1.0   1.000000      36103   

      County Name State  population  
0  Suffolk County    NY     1476601  
1  Suffolk County    NY     1476601  
2  Suffolk County    NY     1476601  
3  Suffolk County    NY     1476601  
4  Suffolk County    NY     1476601  


Was it a clean merge? Check the result for null values.

In [14]:
print(joined_df.isna().sum())

ZIP            0
COUNTY         0
RES_RATIO      0
BUS_RATIO      0
OTH_RATIO      0
TOT_RATIO      0
countyFIPS     0
County Name    0
State          0
population     0
dtype: int64


We find no nulls, so let's move on to the next step and check a few known zip codes. 

We could pick zip codes at random, or we could choose based on our knowledge of our camper households. We happen to know that the vast majority of our campers come from three areas: Charleston (SC), Charlotte (NC), and around Raleigh/Durham/Chapel Hill (NC).

Let's pick a zip code from these places and see if our table returns the right county. 

29409 = Charleston County, SC

28207 = Mecklenburg County, NC

27511 = Wake County, NC

In [15]:
# define a function to match each item
def match(zipcode, county_name):
    
    """ This function matches a zip code with a county name in our joined dataframe. 
    If there's a match, it returns confirmation text. """
    
    located_name = joined_df.loc[joined_df['ZIP'] == zipcode, 'County Name'].values
    if located_name == county_name:
        print('Matched! ' + 'Zip Code ' + zipcode + ' is in ' + county_name + '.')
    else:
        print('No luck. Could not find a match for ' + zipcode)

        
# put our desired matches in a dictionary so we can loop over it       
to_match = {'29409': 'Charleston County', 
            '28207': 'Mecklenburg County', 
            '27511': 'Wake County'}

# match items in the dictionary
for key in to_match:
    match(key, to_match[key])

Matched! Zip Code 29409 is in Charleston County.
Matched! Zip Code 28207 is in Mecklenburg County.
Matched! Zip Code 27511 is in Wake County.


Awesome. Looks like our data match. 

Let's polish the table and get it ready to store in our database. It will be prettier, more parsimonious, and easier to access by SQL if we adjust some of the columns. First let's remind ourselves what the columns are.

In [16]:
print(joined_df.columns)

Index(['ZIP', 'COUNTY', 'RES_RATIO', 'BUS_RATIO', 'OTH_RATIO', 'TOT_RATIO',
       'countyFIPS', 'County Name', 'State', 'population'],
      dtype='object')


Yuck. Such inconsistency! Let's tidy it up.

In [27]:
# drop duplicate county column
single_county_id = joined_df.drop('countyFIPS', axis=1)

# rename columns
remap = {'ZIP':'zipcode', 
         'COUNTY':'county_id', 
         'RES_RATIO':'res_ratio', 
         'BUS_RATIO':'bus_ratio', 
         'OTH_RATIO':'oth_ratio',
         'TOT_RATIO':'tot_ratio', 
         'County Name': 'county_name', 
         'State': 'state', 
         'population': 'county_pop'}

renamed = single_county_id.rename(remap, axis=1)

print(renamed.head())

  zipcode county_id  res_ratio  bus_ratio  oth_ratio  tot_ratio  \
0   00501     36103   0.000000        1.0        0.0   1.000000   
1   06390     36103   0.000000        1.0        1.0   1.000000   
2   11701     36103   0.994713        1.0        1.0   0.995577   
3   11702     36103   1.000000        1.0        1.0   1.000000   
4   11703     36103   1.000000        1.0        1.0   1.000000   

      county_name state  county_pop  
0  Suffolk County    NY     1476601  
1  Suffolk County    NY     1476601  
2  Suffolk County    NY     1476601  
3  Suffolk County    NY     1476601  
4  Suffolk County    NY     1476601  


Looks good! We have one last step before it's ready to send to the database. 

We know we will need a primary key, and zip code seems like a good candidate. Before we can use it as a primary key we need to make sure there are no duplicates. However, it seems like we may run into a problem. We know from earlier research that zip codes can extend across county and even state lines. Let's see if this is represented in the data.

In [56]:
duplicated_test = renamed['zipcode'].duplicated()

print(renamed.shape)
print('=======')
print(renamed[duplicated_test].count())

(46978, 9)
zipcode        13416
county_id      13416
res_ratio      13416
bus_ratio      13416
oth_ratio      13416
tot_ratio      13416
county_name    13416
state          13416
county_pop     13416
dtype: int64


Indeed, we see lots of duplicates. At over 13,000 rows, we have more duplicates than we can eyeball. We need to construct some logic for how to deal with them. 

For our purposes, we want a simple solution that both drops the duplicates and preserves as much accuracy as possible. 

We can achieve this by sorting the dataframe by county population highest to lowest. Then we drop  all duplicates after the first value. This not only gives us the highest chance of correctly guessing the real county, but it also errs on the side of selecting counties (with higher populations) that are more likely to have higher COVID prevalance. 

In [64]:
# sort by population
ordered = renamed.sort_values('county_pop', ascending=False)

# drop duplicates, keeping first value
unique_zips = ordered.drop_duplicates('zipcode', keep='first')

# check output
print(unique_zips.head())

      zipcode county_id  res_ratio  bus_ratio  oth_ratio  tot_ratio  \
32396   60616     17031        1.0        1.0        1.0        1.0   
32288   60153     17031        1.0        1.0        1.0        1.0   
32297   60165     17031        1.0        1.0        1.0        1.0   
32296   60164     17031        1.0        1.0        1.0        1.0   
32295   60163     17031        1.0        1.0        1.0        1.0   

       county_name state  county_pop  
32396  Cook County    IL     5150233  
32288  Cook County    IL     5150233  
32297  Cook County    IL     5150233  
32296  Cook County    IL     5150233  
32295  Cook County    IL     5150233  


Confirm that all zipcode duplicates have been dropped.

In [66]:
duplicated_test_2 = unique_zips['zipcode'].duplicated()

print(unique_zips[duplicated_test_2].count())

zipcode        0
county_id      0
res_ratio      0
bus_ratio      0
oth_ratio      0
tot_ratio      0
county_name    0
state          0
county_pop     0
dtype: int64


Perfect. We now have a primary key for our database table. Finally, let's make one adjustment for simplicity's sake - set the zipcode to the index.

In [67]:
final_joined_df = unique_zips.set_index('zipcode')
print(final_joined_df.head())

        county_id  res_ratio  bus_ratio  oth_ratio  tot_ratio  county_name  \
zipcode                                                                      
60616       17031        1.0        1.0        1.0        1.0  Cook County   
60153       17031        1.0        1.0        1.0        1.0  Cook County   
60165       17031        1.0        1.0        1.0        1.0  Cook County   
60164       17031        1.0        1.0        1.0        1.0  Cook County   
60163       17031        1.0        1.0        1.0        1.0  Cook County   

        state  county_pop  
zipcode                    
60616      IL     5150233  
60153      IL     5150233  
60165      IL     5150233  
60164      IL     5150233  
60163      IL     5150233  


## 4. Adding Table to Database

In [18]:
password = '********'

First we connect to the database. To confirm the connection was made, let's also print the metadata on the table containing household zip codes.

In [68]:
engine = create_engine(f'postgresql://postgres:{password}@localhost:5432/grp_data')

metadata = MetaData()

households = Table('households', metadata, autoload=True, autoload_with=engine)

connection = engine.connect()

print(repr(households))

Table('households', MetaData(bind=None), Column('household_id', INTEGER(), table=<households>, primary_key=True, nullable=False), Column('street', VARCHAR(length=60), table=<households>), Column('city', VARCHAR(length=60), table=<households>), Column('state', VARCHAR(length=60), table=<households>), Column('zipcode', VARCHAR(length=60), table=<households>), Column('country', VARCHAR(length=60), table=<households>), schema=None)


The connection was successful. Now we can create a new table to store our new data in.

In [69]:
counties = Table(
    'counties', metadata, 
    Column('zipcode', String, primary_key=True),
    Column('county_id', String),
    Column('res_ratio', Numeric),
    Column('bus_ratio', Numeric),
    Column('oth_ratio', Numeric),
    Column('tot_ratio', Numeric),
    Column('county_name', String),
    Column('state', String),
    Column('county_pop', Integer),
)

metadata.create_all(engine)

Let's insert our new dataframe into the table.

In [70]:
final_joined_df.to_sql('counties', con=connection, if_exists='append')

As a final check, let's submit a query to the database. If a coherent value is returned, we are done! For our test, let's find the county from the first few rows of households. This will require us to join households to counties.

In [73]:
# build a query statement 
stmt = 'SELECT c.county_name FROM households as h JOIN counties as c USING(zipcode) LIMIT 10'

# execute query
result = pd.read_sql(stmt, connection)

print(result)

          county_name
0  Mecklenburg County
1   Greenville County
2       Fulton County
3      Johnson County
4        Roanoke city
5        Stark County
6     Buncombe County
7        Crisp County
8     Buncombe County
9     Guilford County


Success! Our database now contains a link from zipcodes to county names and county FIPS.

In [74]:
connection.close()